app/src/main/java/net/gsantner/markor/ui/hleditor/Highlighter.java

```diff
@@ -217,24 +345,134 @@ public void draw(@NonNull Canvas canvas, CharSequence text, int start, int end,
-        createTypefaceSpanForMatches(spannable, pattern, "monospace", groupsToMatch);
+    protected final void createMonospaceSpanForMatches(final Pattern pattern, int... groupsToMatch) {
+         createTypefaceSpanForMatches(pattern, "monospace", groupsToMatch);
+    }
+
+    protected final void createColoredUnderlineSpanForMatches(final Pattern pattern, @ColorInt int color, int... groupsToMatch) {
+         createSpanForMatches(pattern, matcher -> new ColorUnderlineSpan(color, null), groupsToMatch);
     }
 
-    protected static void createColoredUnderlineSpanForMatches(Spannable spannable, final Pattern pattern, @ColorInt int color, int... groupsToMatch) {
-        createSpanForMatches(spannable, pattern, (matcher, iM) -> new ColorUnderlineSpan(color, null), groupsToMatch);
+    protected final void createColoredUnderlineSpanForMatches(final Pattern pattern, final Callback.r1<Object, Matcher> creator, int... groupsToMatch) {
+         createSpanForMatches(pattern, creator, groupsToMatch);
     }
 
-    protected static void createColoredUnderlineSpanForMatches(Spannable spannable, final Pattern pattern, final SpanCreator<ParcelableSpan> creator, int... groupsToMatch) {
-        createSpanForMatches(spannable, pattern, creator, groupsToMatch);
+    protected final void createSuperscriptStyleSpanForMatches(final Pattern pattern, final int... groupsToMatch) {
+         createSpanForMatches(pattern, matcher -> new SuperscriptSpan(), groupsToMatch);
     }
 
-    protected static void createSuperscriptStyleSpanForMatches(Spannable spannable, final Pattern pattern, final int... groupsToMatch) {
-        createSpanForMatches(spannable, pattern, (matcher, iM) -> new SuperscriptSpan(), groupsToMatch);
+    protected final void createSubscriptStyleSpanForMatches(final Pattern pattern, final int... groupsToMatch) {
+         createSpanForMatches(pattern, matcher -> new SubscriptSpan(), groupsToMatch);
+    }
+
+    protected final void createTabSpans(final int tabWidth)  {
+        if (tabWidth > 0) {
+            createReplacementSpanForMatches(PATTERN_TAB, tabWidth);
+        }
     }
 
-    protected static void createSubscriptStyleSpanForMatches(Spannable spannable, final Pattern pattern, final int... groupsToMatch) {
-        createSpanForMatches(spannable, pattern, (matcher, iM) -> new SubscriptSpan(), groupsToMatch);
+    protected final void createHighlightLinksSpans() {
+        createSpanForMatches(Patterns.WEB_URL, new HighlightSpan().setForeColor(0xff1ea3fd).setItalic(true).setTextSize(_textSize * 0.85f));
     }
 
-}
+    protected final void createUnderlineHexColorsSpans() {
+        createColoredUnderlineSpanForMatches(HEX_CODE_UNDERLINE_PATTERN, m -> new ColorUnderlineSpan(Color.parseColor(m.group(1)), 3f), 1);
+    }
+
+    // We _do not_ implement UpdateLayout or Parcelable for performance reasons
+    public static class HighlightSpan extends CharacterStyle implements UpdateAppearance, Callback.r1<Object, Matcher> {
```

Comment:
Between this span class and the layout changes, the performance issues have been solved. Now I need to figure out how to solve the jumping and tearing issues in todo.txt

Reply:
guess the jumping is due the border-divider-as-a-span, and them popping up and away?

Reply:
> guess the jumping is due the border-divider-as-a-span, and them popping up and going away? Thats generally a so-far-nice-working-hack on it's own, but don't know other way to insert dividers in a better fashion.

Yes, exactly. It will happen to a lesser degree with larger headings in markdown too. I am working on measuring and restoring scroll position relative to text when we re-highlight.

Reply:
when you dont get far this way, an alternative is to exclude the dividers from the range/screen-height stuff